docs/design/uptime-reputation-calculation.md

```diff
@@ -0,0 +1,122 @@
+
+The process is divided in 2 parts:
+
+1. Selecting _SNs_ for receiving uptime checks.
+1. Periodically check the _SNs_ which has failed the first uptime check until they pass one or get disqualified.
+
+For the first part, we foresee to use the current implemented mechanism with some minor modifications for excluding the _SNs_ which are being checked by the second part.
+
+The second part requires a totally new implementation which should be totally independent of any current satellite service for being able in the near future to run it separately for scalability purpose.
+
+
+## Rationale
+
+The approach described in this document present the following advantages:
+
+1. It's clear why an _SN_ get disqualified due to failed uptime checks.
+1. It has a clear proof for an _SNOs_ why their _SNs_ get disqualified due to failed uptime check in case that they request it before the designed time to clean up the historical failed uptime checks data.
+1. It allows to report to _SN_ at any how far is to be disqualified by failing uptime checks.
+
+And the following disadvantages
+
+1. It requires to store, for some period of time, historical for nodes which have failed at least one uptime check.
+1. It requires more resources than the current implemented approach.
+1. It doesn't calculate the exact time that an _SN_ is offline; once an _SN_ has a failed uptime check the time elapsed between the last failed uptime check and the next succeeded uptime check causes not to have the accurate offline time.
+1. The _"score"_ obtained from uptime check failures, without any further modification, doesn't integrate with the current _SN reputation system_.
+
+While one of our goals is to reduce the number of resources used by the satellite we see that the advantages beat the disadvantages considering that the current system got disabled after some nodes got disqualified unfairly and/or unclear.
+
+On the other hand, we think that the number of resources required by this new system shouldn't be too much and we believe that we can have an implementation that scales up.
+
+## Implementation
+
+### DB
+
+A database is required to store information about _SNs_ uptime check failures. This section presents the new required database schema and a brief explanation what it will store.
+
+Only one database table is required and its schema is the following (pseudo code):
+
+```sql
+CREATE TABLE failed_uptime_checks (
+    node_id       BYTEA NOT NULL,
+    when          TIMESTAMP WITH TIMEZONE NOT NULL, -- When the SN failed the an uptime check until it succeeds a future uptime check
+    back_online   TIMESTAMP WITH TIMEZONE, -- When SN has succeeded an uptime check after the last one failed
+    disqualified  TIMESTAMP WITH TIMEZONE  -- When the SN has been disqualified by any disqualification system (uptimes, audits, etc.)
```

Comment:
Should this be denormalized with some other table? Is this supposed to be the canonical location for if a node is disqualified?

Reply:
No, it isn't the canonical location.

It exists here for being able to run this process separately and not using the DB for sharing state between different services. Perhaps we don't nee this level of independence in such we can query the value from the table which holds it.

Reply:
I expect there will have to be a canonical location for features like manually disqualifying or ... qualifying a node. Indeed, perhaps there are other reasons or factors that go into the decision of disqualification.

Maybe rather than call it "disqualified", call it "stopped_checking"?

Reply:
> Maybe rather than call it "disqualified", call it "stopped_checking"?

I like that.

Reply:
As commented A part of renaming the column, I will mention in the algorithm to check not for _SN_ disqualification but for anything else that we should stop to check the _SN_ because it isn't considered part of the network anymore.

Reply:
I'm holding this until we assess in today's meeting https://github.com/storj/storj/pull/2733#discussion_r312436739